File: apps/ios/Tests/TenantKitTests/CustomerSessionTests.swift
import XCTest

@testable import TenantKit

final class CustomerSessionTests: XCTestCase {
	private func makeSession(
		transport: StubTransport,
		storage: TokenStorage
	) -> CustomerSession {
		CustomerSession(
			client: TenantAPIClient(configuration: makeConfiguration(), transport: transport),
			storage: storage
		)
	}

	func testRestoresStoredSession() async {
		let stored = AuthTokens(accessToken: makeAccessToken(name: "Ada Lovelace"), refreshToken: "refresh-1")
		let session = makeSession(
			transport: StubTransport(),
			storage: InMemoryTokenStorage(tokens: stored)
		)

		let restored = await session.restore()

		XCTAssertEqual(restored?.accessToken, stored.accessToken)
		let isSignedIn = await session.isSignedIn
		XCTAssertTrue(isSignedIn)
		let firstName = await session.customerFirstName
		XCTAssertEqual(firstName, "Ada")
	}

	func testVerifyStoresTokens() async throws {
		let storage = InMemoryTokenStorage()
		let transport = StubTransport(responses: [
			StubTransport.json([
				"success": true,
				"accessToken": makeAccessToken(),
				"refreshToken": "refresh-1",
				"needsName": false,
			]),
		])
		let session = makeSession(transport: transport, storage: storage)

		let result = try await session.verify(phone: "+1 (555) 000-1111", code: " 123456 ")

		XCTAssertEqual(result.needsName, false)
		XCTAssertNotNil(storage.load()?.accessToken)
		let recorded = try XCTUnwrap(transport.lastRequest)
		XCTAssertEqual(recorded.body["phone"] as? String, "+15550001111")
		XCTAssertEqual(recorded.body["code"] as? String, "123456")
	}

	func testRejectsShortPhoneNumbersLocally() async {
		let session = makeSession(transport: StubTransport(), storage: InMemoryTokenStorage())

		do {
			try await session.sendCode(phone: "123")
			XCTFail("Expected a validation error")
		} catch let error as APIError {
			XCTAssertEqual(error.message, "Enter a valid phone number.")
		} catch {
			XCTFail("Unexpected error: \(error)")
		}
	}

	func testAuthorizedRefreshesOnceAfterUnauthorized() async throws {
		let accessToken = makeAccessToken()
		let refreshedToken = makeAccessToken(name: "Ada Lovelace")
		let storage = InMemoryTokenStorage(
			tokens: AuthTokens(accessToken: accessToken, refreshToken: "refresh-1")
		)
		let transport = StubTransport(rawResponses: [
			(status: 401, data: Data(#"{"error":"Invalid or expired token"}"#.utf8)),
			(status: 200, data: Data(#"{"success":true,"accessToken":"\#(refreshedToken)","refreshToken":"refresh-2"}"#.utf8)),
			(status: 200, data: Data(#"{"customer":{"id":"cust_1","name":"Ada Lovelace","needsName":false}}"#.utf8)),
		])
		let session = makeSession(transport: transport, storage: storage)

		let profile = try await session.profile()

		XCTAssertEqual(profile.name, "Ada Lovelace")
		let urls = transport.recorded.map(\.url.path)
		XCTAssertEqual(urls, ["/auth/me", "/auth/refresh", "/auth/me"])
		XCTAssertEqual(storage.load()?.refreshToken, "refresh-2")
	}

	func testAuthorizedSignsOutWhenRefreshFails() async throws {
		let accessToken = makeAccessToken()
		let storage = InMemoryTokenStorage(
			tokens: AuthTokens(accessToken: accessToken, refreshToken: "refresh-1")
		)
		let transport = StubTransport(rawResponses: [
			(status: 401, data: Data(#"{"error":"Invalid or expired token"}"#.utf8)),
			(status: 401, data: Data(#"{"error":"Invalid refresh token"}"#.utf8)),
		])
		let session = makeSession(transport: transport, storage: storage)

		do {
			_ = try await session.profile()
			XCTFail("Expected the session to be cleared")
		} catch let error as APIError {
			XCTAssertTrue(error.isUnauthorized)
		} catch {
			XCTFail("Unexpected error: \(error)")
		}

		XCTAssertNil(storage.load())
		let stillSignedIn = await session.isSignedIn
		XCTAssertFalse(stillSignedIn)
	}

	func testUpdateProfileStoresRotatedTokens() async throws {
		let accessToken = makeAccessToken()
		let rotated = makeAccessToken(name: "Ada Byron")
		let storage = InMemoryTokenStorage(
			tokens: AuthTokens(accessToken: accessToken, refreshToken: "refresh-1")
		)
		let transport = StubTransport(rawResponses: [
			(status: 200, data: Data(#"{"success":true,"accessToken":"\#(rotated)","refreshToken":"refresh-2"}"#.utf8)),
			(status: 200, data: Data(#"{"customer":{"id":"cust_1","name":"Ada Byron","needsName":false}}"#.utf8)),
		])
		let session = makeSession(transport: transport, storage: storage)

		let profile = try await session.updateProfile(name: " Ada Byron ", email: "ada@example.com")

		XCTAssertEqual(profile.name, "Ada Byron")
		XCTAssertEqual(storage.load()?.accessToken, rotated)
		let profileRequest = try XCTUnwrap(transport.recorded.first)
		XCTAssertEqual(profileRequest.body["name"] as? String, "Ada Byron")
		XCTAssertEqual(profileRequest.body["email"] as? String, "ada@example.com")
	}

	func testUpdateProfileValidatesName() async {
		let storage = InMemoryTokenStorage(
			tokens: AuthTokens(accessToken: makeAccessToken(), refreshToken: "refresh-1")
		)
		let session = makeSession(transport: StubTransport(), storage: storage)

		do {
			_ = try await session.updateProfile(name: "A", email: nil)
			XCTFail("Expected a validation error")
		} catch let error as APIError {
			XCTAssertEqual(error.message, "Name is required.")
		} catch {
			XCTFail("Unexpected error: \(error)")
		}
	}

	func testSignOutRevokesRefreshTokenAndClearsStorage() async {
		let storage = InMemoryTokenStorage(
			tokens: AuthTokens(accessToken: makeAccessToken(), refreshToken: "refresh-1")
		)
		let transport = StubTransport(responses: [StubTransport.json(["success": true])])
		let session = makeSession(transport: transport, storage: storage)

		await session.signOut()

		XCTAssertNil(storage.load())
		let recorded = try? XCTUnwrap(transport.lastRequest)
		XCTAssertEqual(recorded?.url.path, "/auth/logout")
		XCTAssertEqual(recorded?.body["refreshToken"] as? String, "refresh-1")
		XCTAssertEqual(recorded?.body["orgId"] as? String, "org_1")
	}

	func testConcurrentUnauthorizedCallsShareOneRefresh() async throws {
		let accessToken = makeAccessToken()
		let refreshed = makeAccessToken(name: "Ada Lovelace")
		let storage = InMemoryTokenStorage(
			tokens: AuthTokens(accessToken: accessToken, refreshToken: "refresh-1")
		)
		// Two 401s (one per caller), then a single refresh, then both retries.
		let transport = StubTransport(rawResponses: [
			(status: 401, data: Data(#"{"error":"Invalid or expired token"}"#.utf8)),
			(status: 401, data: Data(#"{"error":"Invalid or expired token"}"#.utf8)),
			(status: 200, data: Data(#"{"success":true,"accessToken":"\#(refreshed)","refreshToken":"refresh-2"}"#.utf8)),
			(status: 200, data: Data(#"{"customer":{"id":"cust_1","name":"Ada Lovelace"}}"#.utf8)),
			(status: 200, data: Data(#"{"customer":{"id":"cust_1","name":"Ada Lovelace"}}"#.utf8)),
		])
		let session = makeSession(transport: transport, storage: storage)

		async let first = session.profile()
		async let second = session.profile()
		_ = try await (first, second)

		let refreshCalls = transport.recorded.filter { $0.url.path == "/auth/refresh" }
		XCTAssertEqual(refreshCalls.count, 1, "refresh tokens rotate; parallel refreshes would revoke the session")
		XCTAssertEqual(storage.load()?.refreshToken, "refresh-2")
	}

	func testUpdateProfileKeepsRefreshTokenWhenResponseOmitsIt() async throws {
		let storage = InMemoryTokenStorage(
			tokens: AuthTokens(accessToken: makeAccessToken(), refreshToken: "refresh-1")
		)
		let rotated = makeAccessToken(name: "Ada Byron")
		let transport = StubTransport(rawResponses: [
			(status: 200, data: Data(#"{"success":true,"accessToken":"\#(rotated)"}"#.utf8)),
			(status: 200, data: Data(#"{"customer":{"id":"cust_1","name":"Ada Byron"}}"#.utf8)),
		])
		let session = makeSession(transport: transport, storage: storage)

		_ = try await session.updateProfile(name: "Ada Byron", email: nil)

		XCTAssertEqual(storage.load()?.accessToken, rotated)
		XCTAssertEqual(storage.load()?.refreshToken, "refresh-1")
	}

	func testSignOutDiscardsAnInFlightRefresh() async throws {
		let storage = InMemoryTokenStorage(
			tokens: AuthTokens(accessToken: makeAccessToken(), refreshToken: "refresh-1")
		)
		let rotated = makeAccessToken(name: "Ada Lovelace")
		let transport = StubTransport(delayedResponses: [
			(status: 401, data: Data(#"{"error":"Invalid or expired token"}"#.utf8), delayMs: 0),
			// Delayed so sign-out is guaranteed to land while this is in flight.
			(status: 200, data: Data(#"{"success":true,"accessToken":"\#(rotated)","refreshToken":"refresh-2"}"#.utf8), delayMs: 200),
			(status: 200, data: Data(#"{"success":true}"#.utf8), delayMs: 0),
		])
		let session = makeSession(transport: transport, storage: storage)

		async let profile: CustomerProfile? = try? await session.profile()
		try await Task.sleep(nanoseconds: 60_000_000) // let the 401 land and the refresh start
		await session.signOut()
		_ = await profile

		XCTAssertNil(storage.load(), "a refresh that finished after sign-out must not restore the session")
		let isSignedIn = await session.isSignedIn
		XCTAssertFalse(isSignedIn)
	}

	func testTransientRefreshFailureKeepsTheSession() async throws {
		let storage = InMemoryTokenStorage(
			tokens: AuthTokens(accessToken: makeAccessToken(), refreshToken: "refresh-1")
		)
		let transport = StubTransport(rawResponses: [
			(status: 401, data: Data(#"{"error":"Invalid or expired token"}"#.utf8)),
			(status: 500, data: Data(#"{"error":"Internal Server Error"}"#.utf8)),
		])
		let session = makeSession(transport: transport, storage: storage)

		do {
			_ = try await session.profile()
			XCTFail("Expected the server error to surface")
		} catch let error as APIError {
			XCTAssertEqual(error.statusCode, 500)
		} catch {
			XCTFail("Unexpected error: \(error)")
		}

		XCTAssertNotNil(storage.load(), "a 5xx must not sign the customer out")
		XCTAssertEqual(storage.load()?.refreshToken, "refresh-1")
	}

	func testUpdateProfileOmitsEmailWhenNil() async throws {
		let storage = InMemoryTokenStorage(
			tokens: AuthTokens(accessToken: makeAccessToken(), refreshToken: "refresh-1")
		)
		let transport = StubTransport(rawResponses: [
			(status: 200, data: Data(#"{"success":true,"accessToken":"\#(makeAccessToken(name: "Ada"))"}"#.utf8)),
			(status: 200, data: Data(#"{"customer":{"id":"cust_1","name":"Ada","email":"kept@example.com"}}"#.utf8)),
		])
		let session = makeSession(transport: transport, storage: storage)

		_ = try await session.updateProfile(name: "Ada", email: nil)

		let request = try XCTUnwrap(transport.recorded.first)
		XCTAssertNil(request.body["email"], "nil must omit the field so a stored email is not cleared")
		XCTAssertEqual(request.body["name"] as? String, "Ada")
	}

	func testSignedOutSessionRefusesAuthorizedCalls() async {
		let session = makeSession(transport: StubTransport(), storage: InMemoryTokenStorage())

		do {
			_ = try await session.profile()
			XCTFail("Expected an unauthorized error")
		} catch let error as APIError {
			XCTAssertTrue(error.isUnauthorized)
		} catch {
			XCTFail("Unexpected error: \(error)")
		}
	}
}

File: apps/android/app/src/test/kotlin/com/epicstartup/tenant/core/CustomerSessionTests.kt
package com.epicstartup.tenant.core

import com.epicstartup.tenant.core.model.AuthTokens
import com.epicstartup.tenant.core.net.ApiException
import com.epicstartup.tenant.core.net.ApiRequest
import com.epicstartup.tenant.core.net.ApiResponse
import com.epicstartup.tenant.core.net.HttpTransport
import com.epicstartup.tenant.core.net.TenantApiClient
import com.epicstartup.tenant.core.session.CustomerSession
import com.epicstartup.tenant.core.session.InMemoryTokenStorage
import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicReference
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

/** A transport that can hold a response open, to pin down interleavings. */
private class GatedTransport(
	private val gateOn: (ApiRequest) -> Boolean = { true },
	private val handler: (ApiRequest) -> ApiResponse,
) : HttpTransport {
	private val entered = CountDownLatch(1)
	private val release = CountDownLatch(1)
	val requests = CopyOnWriteArrayList<ApiRequest>()

	override fun send(request: ApiRequest): ApiResponse {
		requests += request
		entered.countDown()
		if (gateOn(request)) release.await(5, TimeUnit.SECONDS)
		return handler(request)
	}

	fun awaitFirstRequest(): Boolean = entered.await(5, TimeUnit.SECONDS)

	fun releaseAll() = release.countDown()

	fun refreshRequests(): List<ApiRequest> = requests.filter { it.url.endsWith("/auth/refresh") }
}

class CustomerSessionTest {
	private val accessToken = makeAccessToken()
	private val refreshedToken = makeAccessToken(name = "Jane Doe")

	private fun session(
		storage: InMemoryTokenStorage = InMemoryTokenStorage(),
		handler: (ApiRequest) -> ApiResponse,
	): Pair<CustomerSession, StubTransport> {
		val transport = StubTransport(handler)
		val session = CustomerSession(TenantApiClient(makeConfiguration(), transport), storage)
		return session to transport
	}

	@Test
	fun verifyStoresTheSession() {
		val storage = InMemoryTokenStorage()
		val (session, transport) = session(storage) { _ ->
			jsonResponse(
				200,
				jsonBody("success" to true, "accessToken" to accessToken, "refreshToken" to "refresh_1"),
			)
		}

		val result = session.verify("+1 (555) 000-0000", " 123456 ")

		assertEquals(accessToken, result.accessToken)
		assertTrue(session.isSignedIn)
		assertEquals("org_1", session.tokens?.orgId)
		assertEquals("Jane", session.customerFirstName)
		assertEquals(accessToken, storage.load()?.accessToken)
		// The phone and code are normalized before they are sent, matching tenant-api.
		val body = requestBodyJson(assertNotNull(transport.lastRequest))
		assertEquals("+15550000000", body.getString("phone"))
		assertEquals("123456", body.getString("code"))
	}

	@Test
	fun sendCodeRejectsAnImplausiblePhoneBeforeCallingTheApi() {
		val (session, transport) = session { _ -> jsonResponse(200, jsonBody("success" to true)) }

		val error = assertFailsWith<ApiException> { session.sendCode("12") }

		assertEquals(400, error.statusCode)
		assertTrue(transport.requests.isEmpty())
	}

	@Test
	fun profileUsesTheStoredAccessToken() {
		val (session, transport) = session(InMemoryTokenStorage(AuthTokens(accessToken, "refresh_1"))) { _ ->
			jsonResponse(200, jsonBody("customer" to mapOf("id" to "cus_1", "name" to "Jane Doe")))
		}

		val profile = session.profile()

		assertEquals("Jane Doe", profile.name)
		assertEquals("Bearer $accessToken", assertNotNull(transport.lastRequest).headers["Authorization"])
	}

	@Test
	fun authorizedRefreshesOnceAndRetriesAfterA401() {
		var meCalls = 0
		val (session, transport) = session(InMemoryTokenStorage(AuthTokens(accessToken, "refresh_1"))) { request ->
			when {
				request.url.endsWith("/auth/refresh") -> jsonResponse(
					200,
					jsonBody("accessToken" to refreshedToken, "refreshToken" to "refresh_2"),
				)
				else -> {
					meCalls += 1
					if (meCalls == 1) {
						jsonResponse(401, jsonBody("error" to "Token expired"))
					} else {
						jsonResponse(200, jsonBody("customer" to mapOf("id" to "cus_1", "name" to "Jane")))
					}
				}
			}
		}

		val profile = session.profile()

		assertEquals("Jane", profile.name)
		assertEquals(2, meCalls)
		assertEquals(1, transport.requests.count { it.url.endsWith("/auth/refresh") })
		assertEquals("refresh_2", session.tokens?.refreshToken)
	}

	@Test
	fun authorizedClearsTheSessionWhenTheRefreshTokenIsDead() {
		val storage = InMemoryTokenStorage(AuthTokens(accessToken, "refresh_1"))
		val (session, _) = session(storage) { request ->
			when {
				request.url.endsWith("/auth/refresh") -> jsonResponse(401, jsonBody("error" to "Invalid refresh token"))
				else -> jsonResponse(401, jsonBody("error" to "Token expired"))
			}
		}

		assertFailsWith<ApiException> { session.profile() }

		assertFalse(session.isSignedIn)
		assertNull(storage.load())
	}

	@Test
	fun transientRefreshFailuresKeepTheSession() {
		val storage = InMemoryTokenStorage(AuthTokens(accessToken, "refresh_1"))
		val (session, _) = session(storage) { request ->
			when {
				request.url.endsWith("/auth/refresh") -> jsonResponse(500, jsonBody("error" to "Tenant Database unavailable"))
				else -> jsonResponse(401, jsonBody("error" to "Token expired"))
			}
		}

		assertFailsWith<ApiException> { session.profile() }

		assertTrue(session.isSignedIn)
		assertEquals("refresh_1", storage.load()?.refreshToken)
	}

	@Test
	fun concurrentRefreshesShareOneRotation() {
		val transport = GatedTransport { request ->
			when {
				request.url.endsWith("/auth/refresh") -> jsonResponse(
					200,
					jsonBody("accessToken" to refreshedToken, "refreshToken" to "refresh_2"),
				)
				else -> jsonResponse(200, jsonBody("customer" to mapOf("id" to "cus_1", "name" to "Jane")))
			}
		}
		val session = CustomerSession(
			TenantApiClient(makeConfiguration(), transport),
			InMemoryTokenStorage(AuthTokens(accessToken, "refresh_1")),
		)

		val first = AtomicReference<AuthTokens?>()
		val second = AtomicReference<AuthTokens?>()
		val firstThread = Thread { first.set(session.refresh()) }
		firstThread.start()
		assertTrue(transport.awaitFirstRequest(), "the first refresh should reach the transport")

		val secondThread = Thread { second.set(session.refresh()) }
		secondThread.start()
		// Give the second caller time to join the in-flight refresh before the
		// first response is released.
		Thread.sleep(100)
		transport.releaseAll()
		firstThread.join(5_000)
		secondThread.join(5_000)

		assertEquals(1, transport.refreshRequests().size)
		assertEquals(refreshedToken, first.get()?.accessToken)
		assertEquals(refreshedToken, second.get()?.accessToken)
	}

	@Test
	fun aRefreshThatLandsAfterSignOutCannotRestoreTheSession() {
		val storage = InMemoryTokenStorage(AuthTokens(accessToken, "refresh_1"))
		val transport = GatedTransport { request ->
			when {
				request.url.endsWith("/auth/refresh") -> jsonResponse(
					200,
					jsonBody("accessToken" to refreshedToken, "refreshToken" to "refresh_2"),
				)
				else -> jsonResponse(200, jsonBody("success" to true))
			}
		}
		val session = CustomerSession(TenantApiClient(makeConfiguration(), transport), storage)

		val failure = AtomicReference<Exception?>()
		val refreshThread = Thread {
			try {
				session.refresh()
			} catch (error: Exception) {
				failure.set(error)
			}
		}
		refreshThread.start()
		assertTrue(transport.awaitFirstRequest())

		session.signOut()
		transport.releaseAll()
		refreshThread.join(5_000)

		assertFalse(session.isSignedIn)
		assertNull(storage.load())
		assertNotNull(failure.get(), "the abandoned refresh must not adopt its tokens")
	}

	@Test
	fun aProfileUpdateThatLandsAfterSignOutCannotRestoreTheSession() {
		val storage = InMemoryTokenStorage(AuthTokens(accessToken, "refresh_1"))
		val transport = GatedTransport(gateOn = { it.url.endsWith("/auth/profile") }) { request ->
			when {
				request.url.endsWith("/auth/profile") -> jsonResponse(
					200,
					jsonBody("accessToken" to refreshedToken, "refreshToken" to "refresh_2"),
				)
				else -> jsonResponse(200, jsonBody("success" to true))
			}
		}
		val session = CustomerSession(TenantApiClient(makeConfiguration(), transport), storage)

		val failure = AtomicReference<Exception?>()
		val updateThread = Thread {
			try {
				session.updateProfile("Jane Doe", null)
			} catch (error: Exception) {
				failure.set(error)
			}
		}
		updateThread.start()
		assertTrue(transport.awaitFirstRequest())

		session.signOut()
		transport.releaseAll()
		updateThread.join(5_000)

		assertFalse(session.isSignedIn)
		assertNull(storage.load())
		val error = assertNotNull(failure.get(), "the abandoned update must not adopt its tokens")
		assertTrue((error as ApiException).isUnauthorized)
	}

	@Test
	fun signOutRevokesTheRefreshTokenAndClearsTheSession() {
		val storage = InMemoryTokenStorage(AuthTokens(accessToken, "refresh_1"))
		val (session, transport) = session(storage) { _ -> jsonResponse(200, jsonBody("success" to true)) }

		session.signOut()

		assertFalse(session.isSignedIn)
		assertNull(storage.load())
		val logout = assertNotNull(transport.requests.firstOrNull { it.url.endsWith("/auth/logout") })
		val body = requestBodyJson(logout)
		assertEquals("refresh_1", body.getString("refreshToken"))
		assertEquals("org_1", body.getString("orgId"))
	}

	@Test
	fun clearLocalSessionDropsTheSessionWithoutCallingTheApi() {
		val storage = InMemoryTokenStorage(AuthTokens(accessToken, "refresh_1"))
		val (session, transport) = session(storage) { _ -> jsonResponse(200, jsonBody("success" to true)) }

		session.clearLocalSession()

		assertFalse(session.isSignedIn)
		assertNull(storage.load())
		// A session stored by another tenant must not be revoked through this
		// org's node: that node may be in the other data region.
		assertTrue(transport.requests.isEmpty())
	}

	@Test
	fun coalescedCallersSeeTheOriginalRefreshFailure() {
		val transport = GatedTransport { request ->
			when {
				request.url.endsWith("/auth/refresh") -> jsonResponse(
					401,
					jsonBody("error" to "invalid_grant"),
				)
				else -> jsonResponse(200, jsonBody("success" to true))
			}
		}
		val session = CustomerSession(
			TenantApiClient(makeConfiguration(), transport),
			InMemoryTokenStorage(AuthTokens(accessToken, "refresh_1")),
		)

		val ownerFailure = AtomicReference<Exception?>()
		val joinedFailure = AtomicReference<Exception?>()
		val owner = Thread {
			try {
				session.refresh()
			} catch (error: Exception) {
				ownerFailure.set(error)
			}
		}
		owner.start()
		assertTrue(transport.awaitFirstRequest(), "the first refresh should reach the transport")

		val joined = Thread {
			try {
				session.refresh()
			} catch (error: Exception) {
				joinedFailure.set(error)
			}
		}
		joined.start()
		Thread.sleep(100)
		transport.releaseAll()
		owner.join(5_000)
		joined.join(5_000)

		// Both callers must see the status the owner saw: a flattened transport
		// failure would skip the sign-in prompt in `AppState.loadProfile`.
		assertEquals(401, (assertNotNull(ownerFailure.get()) as ApiException).statusCode)
		val joinedError = assertNotNull(joinedFailure.get()) as ApiException
		assertEquals(401, joinedError.statusCode)
		assertTrue(joinedError.isUnauthorized)
	}

	@Test
	fun updateProfileKeepsTheCurrentRefreshTokenWhenTheResponseOmitsIt() {
		val (session, _) = session(InMemoryTokenStorage(AuthTokens(accessToken, "refresh_1"))) { request ->
			when {
				request.url.endsWith("/auth/profile") -> jsonResponse(200, jsonBody("accessToken" to refreshedToken))
				else -> jsonResponse(200, jsonBody("customer" to mapOf("id" to "cus_1", "name" to "Jane Doe")))
			}
		}

		session.updateProfile(name = "Jane Doe", email = null)

		assertEquals("refresh_1", session.tokens?.refreshToken)
	}

	@Test
	fun updateProfileRejectsATooShortName() {
		val (session, transport) = session(InMemoryTokenStorage(AuthTokens(accessToken, "refresh_1"))) { _ ->
			jsonResponse(200, jsonBody("success" to true))
		}

		val error = assertFailsWith<ApiException> { session.updateProfile(name = " J ", email = null) }

		assertEquals(400, error.statusCode)
		assertTrue(transport.requests.isEmpty())
	}

	@Test
	fun restoreLoadsAPreviouslyStoredSession() {
		val storage = InMemoryTokenStorage(AuthTokens(accessToken, "refresh_1"))
		val session = CustomerSession(TenantApiClient(makeConfiguration(), StubTransport()), storage)

		assertTrue(session.isSignedIn)
		assertEquals(accessToken, session.restore()?.accessToken)
	}
}
